compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java

```diff
@@ -491,7 +514,33 @@ static void generateCodeInternal(ModuleContext moduleContext,
         // Write the bir to the file system
         // This code will execute only if JAR caching is successful
         // TODO: check the filesystem cache and delete if the cache is incomplete (if BIR or JAR is missing)
-        moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
+        if (shouldNotOptimizeModule) {
+            moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
+        }
+    }
+
+    public boolean isWhiteListedModule() {
+        return isObserveModule() || isDriverModule() || isJBallerinaModule();
+    }
+
+    private boolean isObserveModule() {
+        return this.moduleId.moduleName().contains(OBSERVE);
+    }
+
+    private boolean isDriverModule() {
+        return this.moduleId.moduleName().contains(DOT_DRIVER);
+    }
+
+    private boolean isLangLibModule() {
+        return this.moduleDescriptor.moduleCompilationId().toString().startsWith(BALLERINA_LANG);
```

Comment:
This is okay better move this to a util function. [Here is an implementation of isLangLibModule](https://github.com/ballerina-platform/ballerina-lang/blob/471824e0154e97d6586402761d0831d1084fbdd5/misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/completion/util/CommonUtil.java#L117)

Reply:
Relocated the functions to the util class [CodegenOptimizationUtils](https://github.com/ballerina-platform/ballerina-lang/blob/aa8254a61e2f3d88c1fdd944ffe5a5b257ede583/compiler/ballerina-lang/src/main/java/io/ballerina/projects/util/CodegenOptimizationUtils.java#L34) in aa8254a61e2f3d88c1fdd944ffe5a5b257ede583. Will add any reusable functions there in the future.